src/libstd/thread/mod.rs

```diff
@@ -1080,8 +1080,17 @@ impl Thread {
             _ => panic!("inconsistent state in unpark"),
         }
 
-        // Coordinate wakeup through the mutex and a condvar notification
-        let _lock = self.inner.lock.lock().unwrap();
+        // There is a period between when the parked thread sets `state` to
+        // `PARKED` (or last checked `state` in the case of a spurious wake
+        // up) and when it actually waits on `cvar` which is when it can
+        // receive `notify_one`. Fortunately it has `lock` locked at this stage
+        // so we can acquire `lock` to wait until it is ready to receive the
+        // notification.
+        //
+        // Releasing `lock` before the call to `notify_one` means that when the
+        // parked thread wakes it doesn't get woken only to have to wait for us
+        // to release `lock`.
+        self.inner.lock.lock().unwrap();
```

Comment:
Wouldn't that be counterproductive? Since this PR would be literally a NFC then?
But I think this line should be something like `drop(self.inner.lock.lock().unwrap());`

Reply:
`let _ = ...` and `let _x = ...` are very different. The former drops immediately.

But you are right, for extra clarity it is probably better to call `drop` explicitly.